File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys

input = sys.stdin.readline

n, k = map(int, input().split())

weight = [0] * (n + 1)
value = [0] * (n + 1)

dp = [[0] * (k + 1) for _ in range(n + 1)]

for i in range(1, n + 1):
    weight[i], value[i] = map(int, input().split())

for i in range(1, n + 1):
    if weight[i] > k:
        for w in range(1, k + 1):
            dp[i][w] = dp[i - 1][w]
    else :
        for w in range(1, weight[i]):
            dp[i][w] = dp[i - 1][w]
        for w in range(weight[i], k + 1):
            dp[i][w] = max(dp[i - 1][w], dp[i - 1][w - weight[i]] + value[i])

print(dp[n][k])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.swift
let nk = readLine()!.split(separator: " ").map { Int($0)! }
let n = nk[0]
let k = nk[1]

var weight = Array(repeating: 0, count: n + 1)
var value = Array(repeating: 0, count: n + 1)

var dp = Array(repeating: Array(repeating: 0, count: k + 1), count: n + 1)

for i in 1...n {
    let wv = readLine()!.split(separator: " ").map { Int($0)! }
    weight[i] = wv[0]
    value[i] = wv[1]
}

for i in 1...n {
    if weight[i] > k {
        for w in 1...k { dp[i][w] = dp[i - 1][w] }
    } else {
        for w in 1..<weight[i] { dp[i][w] = dp[i - 1][w] }
        for w in weight[i]...k { dp[i][w] = max(dp[i - 1][w], dp[i - 1][w - weight[i]] + value[i]) }
    }
}

print(dp[n][k])
